PDY-1432 Documentation to secure server workloads (#2170)
pivotal/docs-tap: workloads/server.hbs.md

| `port` | The port on which the application is exposed to the rest of the cluster |
| `containerPort` | The port on which the application listens for requests. Defaults to `port` if not set. |
| `name` | A human-readable name for the port. Defaults to `port` if not set. |

> ## <a id="params"></a> Securing `server` workloads

Reviewer: I would call this "Exposing `server` workloads outside the cluster".

Reviewer: Also, you need a unique `id` so people can link to this.

Author: done

> ### Manual configuration

Reviewer 2: ```suggestion
### Manual configuration for HTTP workloads
```

> For http `server` workloads you can expose them securely creating an `Ingress` resource. Taking the `spring-sensors-consumer-web`

Reviewer 2: ```suggestion
For http `server` workloads you can expose them by creating an Ingress resource and using cert-manager to provision TLS signed certificates. Taking the `spring-sensors-consumer-web`
```

workload as an example, create the following `Ingress`:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: spring-sensors-consumer-web
  namespace: <YOUR-DEVELOPER-NAMESPACE>
  annotations:
    cert-manager.io/cluster-issuer: tap-ingress-selfsigned
    ingress.kubernetes.io/force-ssl-redirect: "true"
    kubernetes.io/ingress.class: contour
    kubernetes.io/tls-acme: "true"
spec:
  tls:
    - secretName: spring-sensors-consumer-web
      hosts:
        - "spring-sensors-consumer-web.<INGRESS-DOMAIN>"
  rules:
    - host: "spring-sensors-consumer-web.<INGRESS-DOMAIN>"
      http:
        paths:
          - pathType: Prefix
            path: /
            backend:
              service:
                name: spring-sensors-consumer-web
                port:
                  number: 8080

```

Replace the `<YOUR-DEVELOPER-NAMESPACE>` with your developer namespace and `<INGRESS-DOMAIN>` with the domain name defined

> in `tap-values.yaml` during the installation. Also set the annotation `cert-manager.io/cluster-issuer` to the `ingress_issuer`

Reviewer 2: ```suggestion
in `tap-values.yaml` during the installation. Also, set the annotation `cert-manager.io/cluster-issuer` value to the `shared.ingress_issuer`
```

> set during installation or leave as `tap-ingress-selfsigned` to use the default one. Make sure to update the port exposed by

Reviewer 2: configured during installation or leave it as `tap-ingress-selfsigned` to use the default one. Make sure to update the port exposed by

> set during installation or leave as `tap-ingress-selfsigned` to use the default one. Make sure to update the port exposed by

Reviewer 2: ```suggestion
configured during installation or leave it as `tap-ingress-selfsigned` to use the default one. Make sure to update the port exposed by
```

your `Service` resource, in the previous snippet it is set to `8080`. 

Your `server` workload can be accessed with https like so:
```console
curl -k https://spring-sensors-consumer-web.<INGRESS-DOMAIN>
```

> ### Creating a new workload type that will secure _server_ workloads by default

Reviewer 2: The other thing I was thinking... Other than the manual configuration above, do we want to provide public docs on how to setup this by default? Should we confirm with Shaine or Evan? This is creating another workload type, which is cool, but I was wondering if it is a desirable info to make it available in the public docs.

Author: I will point them to this documentation and ask them about this.

Reviewer: I think it's useful to include the notion in the documentation that you can create additional workloads, and a sample of how to do so.

Reviewer: We might even want to promote this to "defining your own workload type", but let's start here, since the other would require a bunch more authoring work.

Author: I changed this title to be: "Define your own workload type that will expose _server_ workloads outside the cluster"

to be consistent with the previous "Exposing `server` workloads outside the cluster"

>**Note** Make sure you delete the `Ingress` resource previously created.

TAP allows to create new workload types. Start by taking the existing `server-template` `ClusterConfigTemplate` and modify
it to add an `Ingress` resource when this new type of workload is created.

>**Note** This procedure assumes you have `yq` cli installed in your computer. 

1. Save the existing `server-template` in a local file:
```console
kubectl get ClusterConfigTemplate server-template -oyaml > secure-server-template.yaml
```
2. Extract `.spec.ytt` field from this file and create another file
```console

> yq '.spec.ytt' secure-server-template.yaml > spec-ytt.yaml

Reviewer 2: Can you check if your yq version works with yq eval instead? mine doesn't work only specifying the yaml path without using eval. e.g yq eval '.spec.ytt' secure-server-template.yaml > spec-ytt.yaml

Author: I found this: https://mikefarah.gitbook.io/yq/

```
Notice for v4.x versions prior to 4.18.1
Since 4.18.1, yq's 'eval/e' command is the default command and no longers needs to be specified.
Older versions will still need to specify 'eval/e'.
Similarly, '-' is no longer required as a filename to read from STDIN (unless reading from multiple files).
```

this is my version:
```
yq (https://github.com/mikefarah/yq/) version v4.30.5
```

it works with `eval` as well

Reviewer 2: I tried with the yq we use on CI. So, my suggestion is to add "eval" since it works for both, just in case.

3. Edit `spec-ytt.yaml` to add the `Ingress` resource.
   1. The `Ingress` resource snippet looks like this:
```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: #@ data.values.workload.metadata.name
  annotations:
    cert-manager.io/cluster-issuer: tap-ingress-selfsigned
    ingress.kubernetes.io/force-ssl-redirect: "true"
    kubernetes.io/ingress.class: contour
    kubernetes.io/tls-acme: "true"
    kapp.k14s.io/change-rule: "upsert after upserting Services"
  labels: #@ merge_labels({ "app.kubernetes.io/component": "run", "carto.run/workload-name": data.values.workload.metadata.name })
spec:
  tls:
    - secretName: #@ data.values.workload.metadata.name
      hosts:
        - #@ data.values.workload.metadata.name + ".<INGRESS-DOMAIN>"
  rules:
    - host: #@ data.values.workload.metadata.name + ".<INGRESS-DOMAIN>"
      http:
        paths:
          - pathType: Prefix
            path: /
            backend:
              service:
                name: #@ data.values.workload.metadata.name
                port:
                  number: 8080
```
  Additional steps:

    - Replace `INGRESS-DOMAIN` to the ingress domain you set during the installation.

>     - Set the annotation `cert-manager.io/cluster-issuer` to the `ingress_issuer` set during installation or leave as `tap-ingress-selfsigned` to use the default one.

Reviewer 2: ```suggestion
    - Set the annotation `cert-manager.io/cluster-issuer` value to the `shared.ingress_issuer` configured during installation or leave it as `tap-ingress-selfsigned` to use the default one.
```
